### PR TITLE
PSR7 Server and BaseApplication

### PR DESCRIPTION
**Pull request by markstory**
These classes tie the various parts of Cake\Http into something useful. The BaseApplication is used by developers to define their middleware and bootstrap their application. The Server is used to dispatch requests into the Application and emit the responses.

After this is done, there will be a series of pull requests creating PSR7 versions of each core DispatchFilter.

Refs #6960

### CODE REVIEW

File: src/Http/BaseApplication.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+     * Load all the application configuration and bootstrap logic.
+     *
+     * @return void
+     */
+    public function bootstrap()
+    {
+        // Load traditional bootstrap file..
+        require_once $this->configDir . '/bootstrap.php';
+
+        // Load other config files your application needs.
```
Comment (lorenzo) [inline src/Http/BaseApplication.php]: I find odd to have this comment in the core

File: src/Http/BaseApplication.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        $cakeResponse = $this->getDispatcher()->dispatch($cakeRequest, $cakeResponse);
+
+        // Convert the response back into a PSR7 object.
+        return $next($request, ResponseTransformer::toPsr($cakeResponse));
+    }
+
+    /**
+     * Get the ActionDispatcher.
+     *
+     * @return \Spekkoek\ActionDispatcher
```
Comment (lorenzo) [inline src/Http/BaseApplication.php]: wrong class

File: src/Http/Server.php [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        $this->app->bootstrap();
+        $request = $request ?: ServerRequestFactory::fromGlobals();
+        $response = $response ?: new Response();
+
+        $middleware = $this->app->middleware(new MiddlewareStack());
+        if (!($middleware instanceof MiddlewareStack)) {
+            throw new RuntimeException('The application `middleware` method did not return a middleware stack.');
+        }
+        $middleware->push($this->app);
+        $this->dispatchEvent('Server.buildMiddleware', ['middleware' => $middleware]);
```
Comment (lorenzo) [inline src/Http/Server.php]: What is the purpose of this event?

Comment (jippi) [inline src/Http/Server.php]: it will build the middleware? :trollface:

Comment (markstory) [inline src/Http/Server.php]: I figured plugins might want a way to hook into the application middleware being built for zero config style plugins.

File: src/Http/BaseApplication.php (hunk 1 above)
Comment (markstory) [inline src/Http/BaseApplication.php]: Good point.

**Comment by markstory** [comment]
Any other feedback on this?

**Comment by burzum** [comment]
Looks good to me.

**Pull request merged**